### PR TITLE
Adding support for stream write including test coverage

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedITest.scala [hunk 1]
```diff
@@ -178,6 +187,159 @@ class SparkE2EChangeFeedITest
     rowCount shouldEqual 2
   }
 
+  "spark change feed micro batch (incremental)" can "be used to copy data to another container" in {
+    val cosmosEndpoint = TestConfigurations.HOST
+    val cosmosMasterKey = TestConfigurations.MASTER_KEY
+
+    val sourceContainer = cosmosClient.getDatabase(cosmosDatabase).getContainer(cosmosContainer)
+
+    val testId = UUID.randomUUID().toString.replace("-", "")
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedITest.scala:196]: '-' is allowed character for container name. Are you seeing errors in presence of '-'

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedITest.scala [hunk 2]
```diff
@@ -178,6 +187,159 @@ class SparkE2EChangeFeedITest
...
+    var sourceCount: Long = getRecordCountOfContainer(sourceContainer)
+    logInfo(s"RecordCount in source container after first execution: $sourceCount")
+    var targetCount: Long = getRecordCountOfContainer(targetContainer)
+    logInfo(s"RecordCount in target container after first execution: $targetCount")
+
+    processedRecordCount.get() shouldEqual 100L
+    sourceCount shouldEqual 100L
+    sourceCount shouldEqual targetCount
+
+    spark.close()
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedITest.scala:272]: `SparkE2EChangeFeedITest` uses a pre-created spark which is not meant to be closed by the test.

as a pattern if you need to close or re-initialize spark, I would suggest you move the test to a new class which doesn't inherit `Spark` test trait

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadataCache.scala [hunk 3]
```diff
@@ -182,9 +182,25 @@ private object PartitionMetadataCache extends CosmosLoggingTrait {
...
           assertNotNull(lastTotalDocumentSize.get, "lastTotalDocumentSize"),
           assertNotNullOrEmpty(lastContinuationToken.get, "continuationToken")
-        )
+        ))
+      })
+      .onErrorResume((throwable: Throwable) => {
+        throwable match {
+          case cosmosException: CosmosException =>
+            if (tolerateNotFound &&
+              cosmosException.getStatusCode == HttpConstants.StatusCodes.NOTFOUND &&
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadataCache.scala:191]: you can factor out the code to `Exceptions` helper class, we have similar other methods.

https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/Exceptions.scala

**Review by moderakh — APPROVED** [review]
LGTM.

a few minor comments, are not blocking.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosContainerChangeFeedTest.java [hunk 4]
```diff
@@ -844,4 +851,34 @@ private void validateContainerResponse(CosmosContainerProperties containerProper
             .isEqualTo(containerProperties.getId());
 
     }
+
+    static class FlakyTestRetryAnalyzer implements IRetryAnalyzer {
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosContainerChangeFeedTest.java:855]: ACK - fixed

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadataCache.scala (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadataCache.scala:191]: ACK - Done

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedITest.scala (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedITest.scala:272]: Good idea - Thanks. Done

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedITest.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com/azure/cosmos/spark/SparkE2EChangeFeedITest.scala:196]: No - just copy & paste and the fact that I mostly use Guid.ToString("N") in .Net whihc reflects above. I have removed it here in this test case.

**Pull request merged**